File: src/algorithms/go/mergeSort.go
package main

func mergeSort(arr []int) []int {
	if len(arr) > 1 {
		mid := int(len(arr) / 2)
		// left := arr[:mid]
		left := make([]int, len(arr[:mid]))
		copy(left, arr[:mid])
		// right := arr[mid:]
		right := make([]int, len(arr[mid:]))
		copy(right, arr[mid:])

		mergeSort(left)
		mergeSort(right)

		i, j, k := 0, 0, 0

		for i < len(left) && j < len(right) {
			if left[i] <= right[j] {
				arr[k] = left[i]
				i++
			} else {
				arr[k] = right[j]
				j++
			}
			k++
		}

		for i < len(left) {
			arr[k] = left[i]
			i++
			k++
		}

		for j < len(right) {
			arr[k] = right[j]
			j++
			k++
		}
	}
	return arr
}

File: src/algorithms/cpp/mergeSort.cpp
#include <iostream>
#include <vector>

std::vector<int> merge(std::vector<int> &left, std::vector<int> &right)
{
    std::vector<int> merged;
    size_t i = 0, j = 0;

    while (i < left.size() && j < right.size())
    {
        if (left[i] <= right[j])
        {
            merged.push_back(left[i]);
            i++;
        }
        else
        {
            merged.push_back(right[j]);
            j++;
        }
    }

    while (i < left.size())
    {
        merged.push_back(left[i]);
        i++;
    }

    while (j < right.size())
    {
        merged.push_back(right[j]);
        j++;
    }

    return merged;
}

std::vector<int> mergeSort(std::vector<int> &arr)
{
    if (arr.size() <= 1)
    {
        return arr;
    }

    int mid = arr.size() / 2;
    std::vector<int> left(arr.begin(), arr.begin() + mid);
    std::vector<int> right(arr.begin() + mid, arr.end());

    left = mergeSort(left);
    right = mergeSort(right);

    return merge(left, right);
}